rsp_jupyter_extensions/handlers/query.py

```diff
@@ -1,112 +1,288 @@
-        nb = self._get_portal_query_notebook(url)
-        r_qdir = Path("notebooks") / "queries"
-        qdir = Path(os.getenv("HOME", "")) / r_qdir
-        qdir.mkdir(parents=True, exist_ok=True)
-        fname = f"portal_{q_id}.ipynb"
-        r_fpath = r_qdir / fname
-        fpath = qdir / fname
-        fpath.write_text(nb)
-        retval = {
-            "status": 200,
-            "filename": str(fname),
-            "path": str(r_fpath),
-            "url": (
-                os.environ.get("JUPYTERHUB_SERVICE_PREFIX", "")
-                + "/tree/"
-                + str(r_fpath)
-            ),
-            "body": nb,
-        }
-        return json.dumps(retval)
-
-    def _get_portal_query_notebook(self, url: str) -> str:
+    async def _create_query(self, q_value: str, q_type: str) -> str:
+        match q_type:
+            case "tap":
+                return await self._create_tap_query(q_value)
+            case _:
+                raise UnsupportedQueryTypeError(
+                    f"{q_type} is not a supported query type"
+                )
+
+    async def _create_tap_query(self, q_value: str) -> str:
+        # The value should be a URL or a jobref ID
+        this_rsp = os.getenv("EXTERNAL_INSTANCE_URL", "not-an-rsp")
+        if q_value.startswith(this_rsp):
+            # This looks like a URL
+            url = q_value
+            q_id = q_value.split("/")[-1]  # Last component is the jobref ID
+        else:
+            # It's a raw jobref ID
+            url = f"{this_rsp}/api/tap/async/{q_value}"
+            q_id = q_value
+        fname = (
+            Path(os.getenv("JUPYTER_SERVER_ROOT", ""))
+            / "notebooks"
+            / "queries"
+            / f"tap_{q_id}.ipynb"
+        )
+        if fname.is_file():
+            nb=fname.read_text()
+        else:
+            nb = self._get_tap_query_notebook(url)
+        await self.refresh_query_history()  # Opportunistic
+        return _write_notebook_response(nb, fname)
+
+    def _get_ts_query_notebook(
+        self,
+        org: str,
+        repo: str,
+        directory: str,
+        notebook: str,
+        params: dict[str,str]
+    ) -> str:
         """Ask times-square for a rendered notebook."""
+        # Since we know the path we don't have to crawl the base github
+        # response.
+        nb_url = f"github/{org}/{repo}/{directory}/{notebook}"
+
+        # Get the endpoint for the rendered URL
+        obj = self._ts_client.get(nb_url).json()
+        rendered_url = obj["rendered_url"]
+
+        # Retrieve that URL and return the textual response, which is the
+        # string representing the rendered notebook "in unicode", which
+        # means "a string represented in the default encoding".
+        return self._ts_client.get(rendered_url, params=params).text
 
-        # These are all constant for this kind of query
+
+    def _get_nublado_seeds_notebook(
```

Comment:
Extra line for style consistency
```suggestion


    def _get_nublado_seeds_notebook(
```